### PR TITLE
Sort BoardGroup and fix iteration

### LINKED ISSUE #37: Review whether BoardGroups should be ordered or not
There is no inherent order to the boards attached to a robot. We should make a decision whether we want to keep this notion (requiring users to address boards by serial number), or to assign an order (in which case we need to be explicit about what that order is e.g. sort by serial number)

This is relevant to #23.

I'm happy to implement the outcome of this decision in a PR.

### PR DESCRIPTION
**Pull request by kierdavis**
Resolves #37 and #38. See commit messages for more details.

### CODE REVIEW

**Comment by kierdavis** [comment]
I'm not sure what to do about these linting errors, pertaining to single-line docstrings being too long.

```
j5/boards/base.py:105:91: E501 line too long (94 > 90 characters)
tests/boards/test_base.py:254:91: E501 line too long (98 > 90 characters)
```

If I attempt to wrap the text onto multiple lines, the linter thinks that I now have a summary line (missing a period, since it is only the first half of a sentence) and a description line (missing a blank line between it and the preceding summary line).

```
j5/boards/base.py:105:1: D205 1 blank line required between summary line and description
j5/boards/base.py:105:1: D400 First line should end with a period
tests/boards/test_base.py:254:1: D205 1 blank line required between summary line and description
tests/boards/test_base.py:254:1: D400 First line should end with a period
```

Should I just reword the docstring to have a shorter summary sentence?

**Comment by trickeydan** [comment]
I'd be happy with either:
- Add an ignore for `E501` on those lines.
- Split it onto separate linres with a short description at the top e.g:

```
def eat_bees(bee_hive: BeeHive):
    """
    Eat the bees.

    Eat all of the bees in the bee hive whilst also being careful to 
    not allow them to sting my tongue as that would likely be painful
    and make me not want to eat bees in the future.
    """    
    bee_hive.consume()
    bee_hive.consume_more()
```

File: j5/boards/base.py [hunk 1]
```diff
@@ -74,15 +74,16 @@ class BoardGroup:
...
 
         self.update_boards()
 
     def update_boards(self) -> None:
         """Update the boards in this group to see if new boards have been added."""
         self.boards: Dict[str, Board] = {}
-        for board in self.board_class.discover(self._backend):
+        discovered_boards = self.board_class.discover(self._backend)
+        discovered_boards.sort(key=lambda board: board.serial)
+        for board in discovered_boards:
```
Comment (sedders123) [inline j5/boards/base.py]: What is the minimum version of Python we are supporting?

Dictionarys being ordered was introduced to the Python standard in 3.7 and as a side effect to the CPython implementation in 3.6. If we wanted to support Python 3.5 (which is the first to allow syntactical type hints) or non CPython 3.6 we'd have to change `self.boards` to be an `OrderdDict` from `collections`

Comment (kierdavis) [reply j5/boards/base.py]: Good point, we should really specify this somewhere. I've created #79.

**Pull request merged**